test/fuse/linux/read_test.cc

```diff
@@ -0,0 +1,236 @@
+// Copyright 2020 The gVisor Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <errno.h>
+#include <fcntl.h>
+#include <sys/stat.h>
+#include <sys/statfs.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+#include <linux/fuse.h>
+
+#include <string>
+#include <vector>
+
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+#include "test/util/fuse_util.h"
+#include "test/util/test_util.h"
+
+#include "fuse_base.h"
+
+namespace gvisor {
+namespace testing {
+
+namespace {
+
+class ReadTest : public FuseTest {
+  void TearDown() { UnmountFuse(); }
+
+ protected:
+  const std::string test_file_ = "test_file";
+};
+
+TEST_F(ReadTest, ReadWhole) {
+  const std::string test_file_path =
+      JoinPath(mount_point_.path().c_str(), test_file_);
+
+  // Open to get a fd.
+  SetServerInodeLookup(test_file_, S_IFREG | S_IRWXU | S_IRWXG | S_IRWXO);
+  const int open_flag = O_RDWR;
+  struct fuse_out_header out_header = {
+      .len = sizeof(struct fuse_out_header) + sizeof(struct fuse_open_out),
+  };
+  struct fuse_open_out out_payload = {
+      .fh = 1,
+      .open_flags = open_flag,
+  };
+  auto iov_out_open = FuseGenerateIovecs(out_header, out_payload);
+  SetServerResponse(FUSE_OPEN, iov_out_open);
+  FileDescriptor fd =
+      ASSERT_NO_ERRNO_AND_VALUE(Open(test_file_path.c_str(), open_flag));
+
+  // Can replace with SkipServerActualRequest() after mergerd.
+  struct fuse_in_header in_header_open;
+  struct fuse_open_in in_payload_open;
+  auto iov_in_open = FuseGenerateIovecs(in_header_open, in_payload_open);
+  GetServerActualRequest(iov_in_open);
+
+  EXPECT_EQ(in_header_open.len,
+            sizeof(in_header_open) + sizeof(in_payload_open));
+  EXPECT_EQ(in_header_open.opcode, FUSE_OPEN);
+
+  // Prepare for the read.
+  const int n_read = 5;
+  std::string data = std::string(n_read, 'a');
+  struct fuse_out_header read_out_header = {
+      .len = static_cast<uint32_t>(sizeof(struct fuse_out_header) +
+                                   data.size() + 1),
+  };
+  auto iov_out_read = FuseGenerateIovecs(read_out_header, data);
+  SetServerResponse(FUSE_READ, iov_out_read);
+
+  // Read the whole "file".
+  char buf[n_read + 1];
```

Comment:
Use `std::vector<char> buf(n_read)` instead and `buf.data()` in read. Append a '\0' at the end to form a string to compare with `data`.